### DISCUSSION TITLE
new file directory created but it lacks the CreateFile method among its children

### DISCUSSION THREAD

**Discussion by ConorHartin** [FreeOpcUa/opcua-asyncio#1287] [General]
I have asked the question here https://stackoverflow.com/questions/76047913/opc-ua-asyncua-python-new-file-directory-created-but-it-lacks-the-createfile-me but since the SO question I linked to didn't get any responses I figure it's worth asking here as well.
As far as I can tell, a mere Folder (created by the create_folder method) can contain nodes, but not actual files. For that, to support file transfers using ua_file_transfer, I would need a file directory. Unfortunately, manage_nodes does not have a create_file_directory method analogous to create_folder. I attempted to create my own method, but the resulting object still fails to find the CreateFile child when I call the UaDirectory's create_file method on it:
```
    async def add_directory(self, nodeid, bname):
        # This section was mostly copied from manage_nodes.create_folder
        nodeid, qname = _parse_nodeid_qname(nodeid, bname)
        parent = self.server.nodes.objects
        directory_node = make_node(
            self.server,
            await _create_object(parent.server, parent.nodeid,
                                 nodeid, qname, ua.ObjectIds.FileDirectoryType)
        )
        # This was mostly copied from standard_address_space_services
        # It doesn't currently work because self.server doesn't have an add_nodes method
        # standard_address_space_services.create_standard_address_space_Services gets called
        # when the internal server is initialized
        # But this doesn't result in any instance of the FileDirectoryType I create
        # having CreateFile as a child node
        # CreateFile is created as a child of node id 13354, "<FileDirectoryName>"
        # Which in turn is a child of 13353, FileDirectoryType, which I use above
        # attrs = ua.MethodAttributes(
        #     DisplayName=LocalizedText("CreateFile"),
        # )
        # node = ua.AddNodesItem(
        #     BrowseName=QualifiedName("CreateFile", 0),
        #     NodeClass_=NodeClass.Method,
        #     ParentNodeId=NumericNodeId(directory_node.nodeid.Identifier, 0),
        #     ReferenceTypeId=NumericNodeId(47, 0),
        #     NodeAttributes=attrs,
        # )
        # self.server.add_nodes([node])
        return directory_node
```

The FileDirectory type clearly exists in standard_address_space_services and ObjectIds, but I can't find a way to construct one with the expected CreateFile method as a child.

**Comment by schroeder-** [comment]
You can have a look how too instantiate a object see examples/server-instantiate-object.py

**Reply by ConorHartin** [reply]
If I try this line:
```
nodes = await instantiate(self.server.nodes.objects, ua.ObjectIds.FileDirectoryType,
                                transfer_folder_name)
```
It errors out on the line 
`rdesc = await _rdesc_from_node(parent, node_type)`
because the FileDirectoryType was an int rather than an object.
I'm not trying to create a new type, just use the FileDirectoryType already inside the asyncua code. Is there any way to do that or do I have to construct FileDirectoryType as a new type?
I have also tried getting the type this way:
```
standard_address_space_nodes = self.server.iserver.node_mgt_service._aspace._nodes
        file_directory_type_key = next(filter(
            lambda k: k.Identifier == ua.ObjectIds.FileDirectoryType,
            standard_address_space_nodes.keys()))
        file_directory_type = standard_address_space_nodes[file_directory_type_key]
```
But then if I pass in file_directory_type instead of ua.ObjectIds.FileDirectoryType it produces:

> AttributeError: 'NodeData' object has no attribute 'read_attributes'

Evidently the instantiate function expects a Node rather than a NodeData (which was confusingly retrieved from a data structure named "_nodes"), and I had assumed that standard_address_space_services was creating/adding Node objects with calls like ua.AddNodesItem, but I don't know how to get the Node version of the FileDirectoryType.

**Reply by schroeder-** [reply]
`client.get_node(ua.NodeId( ua.ObjectIds.FileDirectoryType))`

**Reply by ConorHartin** [reply]
How can the client even get the node if it doesn't exist because it hasn't been instantiated?

**Reply by schroeder-** [reply]
My bad

`filedir_type_node = server.get_node(ua.NodeId( ua.ObjectIds.FileDirectoryType))`

**Reply by schroeder-** [reply]
Full example:

```
dir_node = self.server.get_node(ua.NodeId( ua.ObjectIds.FileDirectoryType))
nodes = await instantiate(self.server.nodes.objects, dir_node,
                                transfer_folder_name)
```

**Reply by schroeder-** [reply]
Ok the example is wrong in /server-instantiate-object.py this should work:

```
dir_node = self.server.get_node(ua.NodeId( ua.ObjectIds.FileDirectoryType))
new_node_id = ua.NodeId(....)
nodes = await instantiate(self.server.nodes.objects, dir_node, new_node_id, transfer_folder_name)
```

**Reply by ConorHartin** [reply]
I was able to instantiate the node that way. Then I tried to actually transfer a file with these lines:
```
children: list[Node] = await self.server.nodes.objects.get_children()
# ...
transfer_folder_node = next(filter(lambda child: child.nodeid == transfer_folder_id, children))
ua_directory = UaDirectory(transfer_folder_node)
transfer_file = await ua_directory.create_file("test_file.txt", True)
```
Which produces this stack trace (which I've edited to just the relevant bits):
transfer_file = await ua_directory.create_file("test_file.txt", True)
  File "...\asyncua\client\ua_file_transfer.py", line 220, in create_file
    create_file_node = await self._directory_node.get_child("CreateFile")
  File "...\asyncua\common\node.py", line 504, in get_child
    result.StatusCode.check()
  File "...\asyncua\ua\uatypes.py", line 328, in check
    raise UaStatusCodeError(self.value)
asyncua.ua.uaerrors._auto.BadNoMatch: "The requested operation has no match to return."(BadNoMatch)

If I step through the function ViewService._find_elements_in_node I can see that my FileDirectory node has two references in its nodedata. One has the BrowseName QualifiedName(NamespaceIndex=0, Name='Objects'), while the other has QualifiedName(NamespaceIndex=0, Name='FileDirectoryType').

**Reply by MIFBobL** [reply]
Hi Conor, did you ever get a reply to your question? Did you get it working? I've been looking to implement a similar FileDirectoryType system to allow transfer of files from Server to Client. Having not found any evidence of support for this in opcua-asyncio I have reverted to using node-opcua which apparently should support it. However I'd rather implement my server in python if possible as it will be easier to deploy, hence my question.

**Reply by ConorHartin** [reply]
No, I didn't get a reply.

**Reply by schroeder-** [reply]
You have to implement all methods our self. The server has only the type information. Use link_method to handle a method.

**Reply by MIFBobL** [reply]
Thanks schroeder - Thats probably why I couldn't find any worked examples for opcua-asyncio Annex 5 support.  As the FileDirectoryType and all the associated methods are part of the Annex 5 Spec then if the implementation 'supports' the Type information then you would think it would also 'support' all the associated 'standard' methods (custom methods - fair enough) . Having to implement ALL the methods yourself isn't really 'supporting' Annex 5 in my view, rather the current implementation only 'allows support'. I'll get it working with node, as I originally planned,  and then if I have time later I will look at opcua-asyncio again. At least, if I understand correctly, it should be possible...... Guess Conor you need to investigate link_method more to proceed....cheers
